### PR TITLE
OTT-1168: support for traffic shaping based on country code

### PR DESCRIPTION
**Pull request by pm-nilesh-chate**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

**Review by Pubmatic-Dhruv-Sonone — APPROVED** [review]

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+		if _, ok := rCtx.AdapterThrottleMap[partnerConfig[models.BidderCode]]; ok {
+			continue
+		}
+
+		biddingCondition, ok := bidderFilter[partnerConfig[models.BidderCode]]
+		if ok && !evaluateBiddingCondition(data, biddingCondition) {
+			filteredBidders[partnerConfig[models.BidderCode]] = struct{}{}
+			continue
+		}
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/traffic_shapping.go:45]: This should be called from `func (m OpenWrap) handleBeforeValidationHook(`'s loop

m.filterBidder() 

func (m OpenWrap) filterBidder(...) {
if rCtx.AdapterThrottleMap[partnerConfig[models.
..

if m.partnerconfigmap[bidder].countryfilter {
    m.evaluateBiddingCondition(partnerconfigmap[bidder].countryfiltercondition)
}

}

File: modules/pubmatic/openwrap/entrypointhook.go [hunk 2]
```diff
@@ -124,6 +124,7 @@ func (m OpenWrap) handleEntrypointHook(
 			}
 			return 0, err
 		},
+		GeoInfoFetcher: m.geoInfoFetcher,
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/entrypointhook.go:127]: this should be part of `openwrap` not request

File: modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go [hunk 3]
```diff
@@ -83,4 +83,14 @@ type AdConfig struct {
...
+	BidderFilter       *BidderFilter    `json:"bidder-filter,omitempty"`
+}
+
+type BidderFilter struct {
+	Filters []Filter `json:"filter-config,omitempty"`
+}
+
+type Filter struct {
+	Bidders           []string    `json:"bidders,omitempty"`
+	BiddingConditions interface{} `json:"bidding-conditions,omitempty"`
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go:95]: This should be string

File: modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go [hunk 4]
```diff
@@ -83,4 +83,14 @@ type AdConfig struct {
 	Regex              *bool            `json:"regex,omitempty"`
 	UniversalPixel     []UniversalPixel `json:"universalpixel,omitempty"`
 	EnableGAMUrlLookup bool             `json:"enablegamurllookup,omitempty"`
+	BidderFilter       *BidderFilter    `json:"bidder-filter,omitempty"`
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go:86]: do not use - in tag

File: modules/pubmatic/openwrap/cache/gocache/adunit_config.go [hunk 5]
```diff
@@ -19,7 +19,20 @@ func (c *cache) populateCacheWithAdunitConfig(pubID int, profileID, displayVersi
...
+			for _, filter := range defaultAdUnitConfig.BidderFilter.Filters {
+				for _, bidder := range filter.Bidders {
+					bidderfilter[bidder] = filter.BiddingConditions
+				}
+			}
+			if len(bidderfilter) > 0 {
+				cacheKey := key(BidderFilter, pubID, profileID, displayVersion)
+				c.Set(cacheKey, bidderfilter)
+			}
+		}
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/cache/gocache/adunit_config.go:35]: process this data outside this function and save the o/p at partner level in partnerConfigMap

File: modules/pubmatic/openwrap/cache/gocache/gocache.go [hunk 6]
```diff
@@ -21,6 +21,7 @@ const (
 	PubSlotRegex    = "psregex_%d_%d_%d_%d_%s" // slot and its matching regex info at publisher, profile, display version and adapter level
 	PubSlotNameHash = "pslotnamehash_%d"       //publisher slotname hash mapping cache key
 	PubVASTTags     = "pvasttags_%d"           //publisher level vasttags
+	BidderFilter    = "bidderfilter_%d_%d_%d"
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/cache/gocache/gocache.go:24]: this is not requried

File: modules/pubmatic/openwrap/metrics/prometheus/prometheus.go [hunk 7]
```diff
@@ -279,11 +280,28 @@ func newMetrics(cfg *config.PrometheusMetrics, promRegistry *prometheus.Registry
 		"Time taken to serve the vast unwrap request in Milliseconds at wrapper count level", []string{pubIdLabel, wrapperCountLabel},
 		[]float64{50, 100, 150, 200})
 
+	metrics.geoDBInitStatus = newGauge(cfg, promRegistry,
+		"geodb_status",
+		"An indicator to identify the GeoDB database's state of failure. 1 indicates failure and 0 indicates healthy.",
+		[]string{dcNameLabel, nodeNameLabel, podNameLabel})
+
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/metrics/prometheus/prometheus.go:287]: This is service startup data, why do we need stats for this. Can we either have retry mechanism or block service startup.
Ideally we should have ip-to-geo lookup status stats. Which will give both the data (if service is up or not and count of success/failure)

File: modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go [hunk 8]
```diff
@@ -25,6 +25,24 @@ var testAdunitConfig = &adunitconfig.AdUnitConfig{
 	Regex:         true,
 	Config: map[string]*adunitconfig.AdConfig{
 		"default": {
+			BidderFilter: &adunitconfig.BidderFilter{
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go:28]: update ut in db layer to include bidderfilter data

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/prebid/prebid-server/v2/modules/pubmatic/openwrap/cache"
+	"github.com/prebid/prebid-server/v2/modules/pubmatic/openwrap/geodb"
+	"github.com/prebid/prebid-server/v2/modules/pubmatic/openwrap/models"
+)
+
+const (
+	keycountry = "country"
+)
+
+func getFilteredBidders(rCtx models.RequestCtx, bidRequest *openrtb2.BidRequest, c cache.Cache) (map[string]struct{}, bool) {
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Done

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+		if bidRequest.Device.IP != "" {
+			ip = bidRequest.Device.IP
+		} else {
+			ip = bidRequest.Device.IPv6
+		}
+	}
+
+	if ip == "" {
+		ip = rCtx.IP
+	}
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Done

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+	if ip != "" {
+		country, err := getCountryFromIP(rCtx.GeoInfoFetcher, ip)
+		if err != nil {
+			glog.Errorf("type:[geo_fetch_failed] ip:[%v] error:[%v]", ip, err)
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Done

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+			ip = bidRequest.Device.IPv6
+		}
+	}
+
+	if ip == "" {
+		ip = rCtx.IP
+	}
+
+	if ip != "" {
+		country, err := getCountryFromIP(rCtx.GeoInfoFetcher, ip)
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Done

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Done

File: modules/pubmatic/openwrap/traffic_shapping.go [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+			continue
+		}
+		allPartnersFilteredFlag = false
+	}
+
+	return filteredBidders, allPartnersFilteredFlag
+}
+
+func generateEvaluationData(rCtx models.RequestCtx, BidRequest *openrtb2.BidRequest) map[string]interface{} {
+	data := map[string]interface{}{}
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/traffic_shapping.go]: Switch to string based implementation

File: modules/pubmatic/openwrap/traffic_shapping.go (hunk 1 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/traffic_shapping.go]: This is at profile level and not impression level, kept the implementation as it.

File: modules/pubmatic/openwrap/entrypointhook.go (hunk 2 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/entrypointhook.go:127]: Done

File: modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go (hunk 3 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go]: Done

File: modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go (hunk 8 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go:28]: Done

File: modules/pubmatic/openwrap/metrics/prometheus/prometheus.go (hunk 7 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/metrics/prometheus/prometheus.go:287]: @ashishshinde-pubm @Ankit-Pinge  Can you please provide your comments here?

File: modules/pubmatic/openwrap/cache/gocache/gocache.go (hunk 6 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/cache/gocache/gocache.go:24]: Removed

File: modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go (hunk 4 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go]: Done

File: modules/pubmatic/openwrap/cache/gocache/adunit_config.go (hunk 5 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/cache/gocache/adunit_config.go]: Done

File: modules/pubmatic/openwrap/cache/gocache/adunit_config.go [hunk 15]
```diff
@@ -19,7 +19,6 @@ func (c *cache) populateCacheWithAdunitConfig(pubID int, profileID, displayVersi
 			v.UniversalPixel = validUPixels(v.UniversalPixel)
 			caseFoldConfigMap[strings.ToLower(k)] = v
 		}
-		adunitConfig.Config = caseFoldConfigMap
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/cache/gocache/adunit_config.go:22]: revert this

Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/cache/gocache/adunit_config.go]: Done

**Review by ashishshinde-pubm — APPROVED** [review]

**Pull request merged**